s3proxy/src/main.rs

```diff
@@ -0,0 +1,678 @@
+        .with_graceful_shutdown(async move { cancel.cancelled().await })
+        .await?;
+    Ok(())
+}
+
+// simplified version of utils::auth::JwtAuth
+pub struct JwtAuth {
+    decoding_key: DecodingKey,
+    validation: Validation,
+}
+
+const VALIDATION_ALGO: jsonwebtoken::Algorithm = jsonwebtoken::Algorithm::EdDSA;
+impl JwtAuth {
+    pub fn new(key: &[u8]) -> Result<Self> {
+        let decoding_key = DecodingKey::from_ed_pem(key)?;
+        let mut validation = Validation::new(VALIDATION_ALGO);
+        validation.required_spec_claims = [].into();
+        Ok(Self {
+            decoding_key,
+            validation,
+        })
+    }
+
+    pub fn decode(&self, token: &str) -> Result<Claims> {
+        Ok(jsonwebtoken::decode(token, &self.decoding_key, &self.validation).map(|t| t.claims)?)
+    }
+}
+
+async fn check_storage_permissions(
+    client: &GenericRemoteStorage,
+    cancel: CancellationToken,
+) -> Result<()> {
+    debug!("Storage permissions check");
+
+    let now = std::time::SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)?
+        .as_secs()
+        .to_string();
+
+    // TODO what if multiple instances try to write a file?
+    // random delay before writing? Instances will be ~3 per region + dedicated for computes
+    let path = RemotePath::from_string(&format!("write_access_{now}"))?;
+    let body_str = format!("{now}");
+    let stream = Body::from(body_str.clone()).into_data_stream();
+    let stream = body_stream_to_sync_stream(stream);
+
+    debug!(%path, "uploading");
+    client
+        .upload(stream, body_str.len(), &path, None, &cancel)
+        .await
+        .context(format!("uploading {path} to test permissions"))?;
+
+    debug!(%path, "downloading");
+    let download_opts = DownloadOpts {
+        kind: remote_storage::DownloadKind::Small,
+        ..Default::default()
+    };
+    let mut stream = client
+        .download(&path, &download_opts, &cancel)
+        .await
+        .context(format!("downloading {path} to test permissions"))?
+        .download_stream;
+    let mut out = Vec::new();
+    use futures::stream::StreamExt;
+    while let Some(res) = stream.next().await {
+        out.extend_from_slice(&res?[..]);
+    }
+    let read_back = String::from_utf8(out)?;
+    if body_str != read_back {
+        bail!("{} (original) != {} (read back)", body_str, read_back)
+    }
+
+    debug!(%path, "removing");
+    client.delete(&path, &cancel).await
+}
+
+fn body_stream_to_sync_stream(
+    stream: BodyDataStream,
+) -> impl futures::Stream<Item = std::io::Result<axum::body::Bytes>> {
+    use futures::stream::TryStreamExt;
+    sync_wrapper::SyncStream::new(stream)
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e.to_string()))
+}
+
+fn app() -> Router {
+    let router = Router::new().route(
+        "/{tenant_id}/{timeline_id}/{endpoint_id}/{key}",
+        get(get_key).put(set_key).delete(delete_key),
```

Comment:
I think more of an API outlined here https://github.com/neondatabase/neon/pull/9661#discussion_r1987750964

To me, 'key' doesn't much sense here, it should be a generic path, i.e. relative path like 'prewarm.data' or 'stats/pg_stat_statements.data' and so on. I haven't checked the code closely yet, maybe it already works this way, but then the naming is confusing.

This service should be completely agnostic of what comes after the initial 'auth' prefix with tenant/timeline/endpoint

Reply:
Fixed with wildcard path